### PR TITLE
SWT: Make circular padding wrap more than once if needed

### PR DESCRIPTION
**Pull request by NiclasPi**
Some combinations of wavelets and number of decomposition steps cause the following `RuntimeError`: "Padding value causes wrapping around more than once".
If the padding values are greater than the input length, `torch.nn.functional.pad` does not wrap around more than once in circluar mode. We need to manually wrap around more than once, if needed. See feature request [https://github.com/pytorch/pytorch/issues/57911](url) in PyTorch.

### CODE REVIEW

File: src/ptwt/_stationary_transform.py [hunk 1]
```diff
@@ -14,6 +15,22 @@
 )
 
 
+def _wrap_padding(x: torch.Tensor, paddings: Sequence[int]) -> torch.Tensor:
+    """Pads a tensor in circular mode more than once if needed"""
+    s = x.shape[-1]  # TODO rename s to something more meaningful
+
+    # TODO write what the regular case is
```
Comment (cthoyt) [inline src/ptwt/_stationary_transform.py:22]: @NiclasPi would you please address the three TODOs I added in the code for improving internal documentation?

File: src/ptwt/_stationary_transform.py [hunk 2]
```diff
@@ -14,6 +15,22 @@
...
+    if not any(padding > s for padding in paddings):
+        return F.pad(x, paddings, mode="circular")
+
+    # TODO what is this doing
+    while any(padding > 0 for padding in paddings):
+        x = F.pad(x, [min(s, padding) for padding in paddings], mode="circular")
+        paddings = [max(padding - s, 0) for padding in paddings]
+
+    return x
```
Comment (cthoyt) [inline src/ptwt/_stationary_transform.py:31]: this function needs its own direct unit tests now, too

**Comment by NiclasPi** [comment]
Thank you for the review @cthoyt ! I added explainig comments and some tests comparing the function with numpys wrap padding.

**Comment by v0lta** [comment]
For posterity: docs to checkout forked code: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/checking-out-pull-requests-locally#modifying-an-inactive-pull-request-locally

**Comment by v0lta** [comment]
```
FAILED tests/test_swt.py::test_swt_1d[db1-size0-1] - assert False
FAILED tests/test_swt.py::test_swt_1d[db1-size0-2] - assert False
FAILED tests/test_swt.py::test_swt_1d[db1-size0-None] - assert False
FAILED tests/test_swt.py::test_swt_1d[db2-size0-1] - assert False
FAILED tests/test_swt.py::test_swt_1d[db2-size0-2] - assert False
FAILED tests/test_swt.py::test_swt_1d[db2-size0-None] - assert False
FAILED tests/test_swt.py::test_swt_1d[db3-size0-1] - assert False
FAILED tests/test_swt.py::test_swt_1d[db3-size0-2] - assert False
FAILED tests/test_swt.py::test_swt_1d[db3-size0-None] - assert False
FAILED tests/test_swt.py::test_swt_1d[db4-size0-1] - assert False
FAILED tests/test_swt.py::test_swt_1d[db4-size0-2] - assert False
FAILED tests/test_swt.py::test_swt_1d[db4-size0-None] - assert False
```
a partial log from `nox -s test`

**Comment by cthoyt** [comment]
@v0lta we need to update this line https://github.com/v0lta/PyTorch-Wavelet-Toolbox/blob/1b75acdbe7c9cbfb526fd1a7c19ad260b935c6ec/.github/workflows/tests.yml#L3 to make tests run on external PRs, it should look like this https://github.com/biopragmatics/curies/blob/470f71a69264c17260823d485de113695076a38f/.github/workflows/tests.yml#L3-L7

**Comment by v0lta** [comment]
Done in the main branch.

**Comment by cthoyt** [comment]
@v0lta @NiclasPi the tests are now showing up properly in the PR

**Comment by v0lta** [comment]
TODO: Port shape fold-unfolding code to swt, properly.

**Comment by NiclasPi** [comment]
Sorry, I forgot to mention that I added test cases that weren't covered yet. Specifically, since I am working with 1-dimensional signals, I added cases with `shape = (N,)`. The code base currently only tests 1-dimensional signals with `shape = (1, N)`.
After closer inspection it turned out to be a problem in the folding-unfolding helper functions. These should also work with `shape = (N,)`. @v0lta is going to look into this issue and will fix it. Afterwards, the failing tests for the SWT should pass.

**Comment by v0lta** [comment]
Thanks @NiclasPi . I have a fix in https://github.com/v0lta/PyTorch-Wavelet-Toolbox/tree/fix-padding which starts to address this. We currently get (1, T) outputs shapes for (N,) inputs across the entire toolbox. With N, the input measurements and T for the dimension at some transform level. This is not what we want since the Toolbox should always respect the uses choice of input dimension. I will fix this across the board.
@cthoyt how did you `[Merge branch 'main' into pr/84]`. My changes ended up in https://github.com/v0lta/PyTorch-Wavelet-Toolbox/tree/fix-padding , unfortunately. I think I should know this but here we are ;-).

**Comment by v0lta** [comment]
It's not a catastrophic problem users can currently fix this by running out.squeeze() themselves, but since we want to be pywt-compatible the shapes should be identical, so we'll fix this.

**Comment by cthoyt** [comment]
you need to add the remote (see https://docs.github.com/en/get-started/getting-started-with-git/managing-remote-repositories)

```
git remote add origin https://github.com/NiclasPi/PyTorch-Wavelet-Toolbox.git
git fetch --all
``

then you can switch to the right fix-padding branch

**Comment by v0lta** [comment]
Hi Team, I removed the `_conv_transpose_dedilate`; the code is simpler and faster now. I also made the SWT module public by removing the underscore. From my point of view, this PR is ready for a merge.
